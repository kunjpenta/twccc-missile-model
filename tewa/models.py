# tewa/models.py

from __future__ import annotations

import uuid
from typing import Optional

from django.conf import settings
from django.core.exceptions import ValidationError
from django.core.validators import MaxValueValidator, MinValueValidator
from django.db import models
from django.utils import timezone
from django.utils.dateparse import parse_datetime


# ---------- Base ----------
class TimeStamped(models.Model):
    created_at = models.DateTimeField(auto_now_add=True, db_index=True)
    updated_at = models.DateTimeField(auto_now=True)

    class Meta:
        abstract = True


# ---------- Validators ----------
lat_validator = [MinValueValidator(-90.0), MaxValueValidator(90.0)]
lon_validator = [MinValueValidator(-180.0), MaxValueValidator(180.0)]
heading_validator = [MinValueValidator(0.0), MaxValueValidator(360.0)]
nonneg_validator = [MinValueValidator(0.0)]  # for speed, alt if you prefer >=0


# tewa/models.py


# Reuse these if you already defined them globally
lat_validator = [MinValueValidator(-90.0),  MaxValueValidator(90.0)]
lon_validator = [MinValueValidator(-180.0), MaxValueValidator(180.0)]


# assume these validators already exist in your module
# from .validators import lat_validator, lon_validator


class Scenario(models.Model):
    id = models.BigAutoField(primary_key=True)

    name = models.CharField(max_length=128, unique=True)
    description = models.TextField(blank=True, null=True)

    start_time = models.DateTimeField(blank=True, null=True, db_index=True)
    end_time = models.DateTimeField(blank=True, null=True, db_index=True)
    notes = models.TextField(blank=True, null=True)

    created_at = models.DateTimeField(auto_now_add=True, db_index=True)
    updated_at = models.DateTimeField(auto_now=True)

    def __str__(self) -> str:
        return self.name


class DefendedAsset(models.Model):
    id = models.BigAutoField(primary_key=True)

    # Make scenario optional so tests can create DA without one
    scenario = models.ForeignKey(
        "Scenario",
        on_delete=models.CASCADE,
        related_name="defended_assets",
        null=True,
        blank=True,
        db_index=True,
    )

    name = models.CharField(max_length=255, db_index=True)
    lat = models.FloatField(validators=lat_validator)
    lon = models.FloatField(validators=lon_validator)
    radius_km = models.FloatField(
        validators=[MinValueValidator(0.001), MaxValueValidator(1000.0)]
    )

    created_at = models.DateTimeField(auto_now_add=True, db_index=True)
    updated_at = models.DateTimeField(auto_now=True)

    def clean(self):
        if self.radius_km is not None:
            try:
                val = float(self.radius_km)
            except (TypeError, ValueError):
                raise ValidationError("Radius_km must be a number.")
            if not (0 < val <= 1000):
                raise ValidationError(
                    "Radius_km must be between 0 and 1000 km.")

    def __str__(self) -> str:
        # radius without trailing .0; lat/lon printed as given
        r = float(self.radius_km)
        r_str = str(int(r)) if r.is_integer() else str(r)
        return f"{self.name} at {self.lat}, {self.lon} with radius {r_str} km"

    class Meta:
        indexes = [
            models.Index(fields=["scenario", "name"], name="idx_da_scn_name"),
            models.Index(fields=["lat", "lon"], name="idx_da_lat_lon"),
        ]
        unique_together = [("scenario", "name")]


class Track(TimeStamped):
    scenario = models.ForeignKey(
        Scenario, on_delete=models.CASCADE, related_name="tracks",
        null=False, blank=False, db_index=True
    )
    track_id = models.CharField(max_length=64, db_index=True)

    lat = models.FloatField(validators=lat_validator)
    lon = models.FloatField(validators=lon_validator)
    alt_m = models.FloatField()  # add nonneg validator if you require >= 0
    speed_mps = models.FloatField(validators=nonneg_validator)
    heading_deg = models.FloatField(validators=heading_validator)

    def __str__(self) -> str:
        return f"{self.track_id} @ {self.scenario.name}"

    class Meta:
        unique_together = [("scenario", "track_id")]
        indexes = [
            models.Index(fields=["scenario", "track_id"],
                         name="idx_track_scn_id"),
            models.Index(fields=["lat", "lon"], name="idx_track_lat_lon"),
            models.Index(fields=["scenario"], name="idx_track_scn"),
        ]


# ---------- TrackSample ----------
class TrackSample(TimeStamped):
    track = models.ForeignKey(
        Track, on_delete=models.CASCADE, related_name="samples", db_index=True
    )
    t = models.DateTimeField(db_index=True)  # sample timestamp
    lat = models.FloatField(validators=lat_validator)
    lon = models.FloatField(validators=lon_validator)
    alt_m = models.FloatField()
    speed_mps = models.FloatField(validators=nonneg_validator)
    heading_deg = models.FloatField(validators=heading_validator)

    def __str__(self) -> str:
        return f"{self.track.track_id} @ {self.t.isoformat()}"

    class Meta:
        ordering = ["t"]
        unique_together = [("track", "t")]
        indexes = [
            models.Index(fields=["track", "t"], name="idx_sample_track_t"),
            models.Index(fields=["lat", "lon"], name="idx_sample_lat_lon"),
        ]


# ---------- ThreatScore ----------
class ThreatScore(TimeStamped):
    """
    Output of the deterministic threat evaluation for a (Track, DA) at a given instant.
    """
    scenario = models.ForeignKey(
        Scenario, on_delete=models.CASCADE, related_name="threat_scores",
        null=False, blank=False, db_index=True
    )
    track = models.ForeignKey(
        Track, on_delete=models.CASCADE, related_name="threat_scores", db_index=True
    )
    da = models.ForeignKey(
        DefendedAsset, on_delete=models.CASCADE, related_name="threat_scores", db_index=True
    )

    # Optional batch tag to group rows produced by a single compute call
    batch_id = models.UUIDField(
        default=uuid.uuid4,  # <-- now uuid is used
        editable=False,
        db_index=True,
        null=False,  # if you want it mandatory
        blank=False,  # if you want it optional
    )

    # Raw components (nullable if not computed)
    cpa_km = models.FloatField(null=True, blank=True)
    tcpa_s = models.FloatField(null=True, blank=True)
    tdb_km = models.FloatField(null=True, blank=True)
    twrp_s = models.FloatField(null=True, blank=True)

    # Final score
    score = models.FloatField(null=True, blank=True)

    # When the compute considered the state
    computed_at = models.DateTimeField(auto_now_add=True, db_index=True)

    def __str__(self) -> str:
        return f"ThreatScore[{self.scenario.name} | {self.track.track_id} â†’ {self.da.name}]"

    class Meta:
        indexes = [
            models.Index(fields=["scenario", "da",
                         "-computed_at"], name="idx_ts_scn_da_cmp"),
            models.Index(fields=["scenario", "track",
                         "-computed_at"], name="idx_ts_scn_trk_cmp"),
            models.Index(fields=["batch_id"], name="idx_ts_batch"),
        ]
        # No unique_together on computed_at to avoid collisions

# ---------- ModelParams ----------


class ModelParams(models.Model):
    """
    Parameter set (weights, normalizers) used by deterministic threat models.
    Scope: per Scenario (default one-to-one).
    """
    scenario = models.OneToOneField(
        Scenario, on_delete=models.CASCADE, related_name='params'
    )

    w_cpa = models.FloatField(default=0.25)
    w_tcpa = models.FloatField(default=0.25)
    w_tdb = models.FloatField(default=0.25)
    w_twrp = models.FloatField(default=0.25)

    cpa_scale_km = models.FloatField(default=20.0)
    tcpa_scale_s = models.FloatField(default=120.0)
    tdb_scale_km = models.FloatField(default=30.0)
    twrp_scale_s = models.FloatField(default=120.0)

    clamp_0_1 = models.BooleanField(default=True)

    created_at = models.DateTimeField(default=timezone.now, db_index=True)
    updated_at = models.DateTimeField(auto_now=True)
    R_W_m = models.PositiveIntegerField(
        default=25000, help_text="Weapon range (m)")
    R_DA_m = models.PositiveIntegerField(
        default=8000, help_text="DA radius (m)")
    tick_s = models.FloatField(default=1.0, validators=[
                               MinValueValidator(1e-6)], help_text="Tick rate (s)")

    # Weights (UI enforces sum=1)
    w_cpa = models.FloatField(default=0.35, validators=[
                              MinValueValidator(0.0), MaxValueValidator(1.0)])
    w_tcpa = models.FloatField(default=0.25, validators=[
                               MinValueValidator(0.0), MaxValueValidator(1.0)])
    w_tdb = models.FloatField(default=0.20, validators=[
                              MinValueValidator(0.0), MaxValueValidator(1.0)])
    w_twrp = models.FloatField(default=0.20, validators=[
                               MinValueValidator(0.0), MaxValueValidator(1.0)])

    # Optional sigmas
    sigma_cpa = models.FloatField(null=True, blank=True)
    sigma_tcpa = models.FloatField(null=True, blank=True)
    sigma_tdb = models.FloatField(null=True, blank=True)
    sigma_twrp = models.FloatField(null=True, blank=True)

    # Optional audit
    updated_by = models.ForeignKey(
        getattr(settings, "AUTH_USER_MODEL", "auth.User"),
        null=True, blank=True, on_delete=models.SET_NULL, related_name="updated_params"
    )

    def __str__(self):
        return f"Params for {self.scenario.name}"


# tewa/models.py (snippets)
# tewa/models.py

# imports at top of file


class ThreatScoreQuerySet(models.QuerySet):
    def for_key(self, *, scenario_id: int, da_id: int, track_id: int) -> "ThreatScoreQuerySet":
        return self.filter(scenario_id=scenario_id, da_id=da_id, track_id=track_id)

    def at_or_latest(self, at_iso: Optional[str] = None):
        qs = self
        if at_iso:
            at_dt = parse_datetime(at_iso)
            if at_dt:
                qs = qs.filter(computed_at__lte=at_dt)
        return qs.order_by("-computed_at", "-id").first()


class ThreatScoreManager(models.Manager):
    def get_queryset(self) -> ThreatScoreQuerySet:  # type: ignore[override]
        return ThreatScoreQuerySet(self.model, using=self._db)

    def latest_for(
        self, *, scenario_id: int, da_id: int, track_id: int, at_iso: Optional[str] = None
    ):
        return (
            self.get_queryset()
            .for_key(scenario_id=scenario_id, da_id=da_id, track_id=track_id)
            .at_or_latest(at_iso=at_iso)
        )


# tewa/models.py (inside ModelParams)

# --- ranges & timing ---
R_W_m = models.PositiveIntegerField(
    default=25000, help_text="Weapon range (m)")
R_DA_m = models.PositiveIntegerField(default=8000, help_text="DA radius (m)")
tick_s = models.FloatField(default=1.0, validators=[
                           MinValueValidator(1e-6)], help_text="Tick rate (s)")

# --- weights (must sum to 1.0 at the form/API layer) ---
w_cpa = models.FloatField(default=0.35, validators=[
                          MinValueValidator(0.0), MaxValueValidator(1.0)])
w_tcpa = models.FloatField(default=0.25, validators=[
                           MinValueValidator(0.0), MaxValueValidator(1.0)])
w_tdb = models.FloatField(default=0.20, validators=[
                          MinValueValidator(0.0), MaxValueValidator(1.0)])
w_twrp = models.FloatField(default=0.20, validators=[
                           MinValueValidator(0.0), MaxValueValidator(1.0)])

# --- sigmas (optional) ---
sigma_cpa = models.FloatField(null=True, blank=True)
sigma_tcpa = models.FloatField(null=True, blank=True)
sigma_tdb = models.FloatField(null=True, blank=True)
sigma_twrp = models.FloatField(null=True, blank=True)

# --- audit (optional if you already have updated_at/created_at via mixin) ---
updated_by = models.ForeignKey(
    getattr(settings, "AUTH_USER_MODEL", "auth.User"),
    null=True, blank=True, on_delete=models.SET_NULL, related_name="updated_params"
)
