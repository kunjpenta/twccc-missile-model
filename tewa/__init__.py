default_app_config = "tewa.apps.TewaConfig"
