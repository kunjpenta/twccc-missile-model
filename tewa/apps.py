from django.apps import AppConfig


class TewaConfig(AppConfig):
    default_auto_field = "django.db.models.BigAutoField"
    name = "tewa"
