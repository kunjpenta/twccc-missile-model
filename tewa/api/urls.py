# tewa/api/urls.py
from django.http import JsonResponse
from django.urls import path
from django.utils.translation import activate
from rest_framework.routers import DefaultRouter

from tewa.api import views_assets_tracks

from . import views
from .views import (
    ScenarioParamsView,
    score_breakdown,  # <- function view
    score_history_png_view,  # <- function view
)

activate("en")


router = DefaultRouter()
router.register(r"scenario",     views.ScenarioViewSet,
                basename="scenario")
router.register(r"da",           views.DefendedAssetViewSet, basename="da")
router.register(r"track",        views.TrackViewSet,         basename="track")
router.register(r"tracksample",  views.TrackSampleViewSet,
                basename="tracksample")
router.register(r"threatscore",  views.ThreatScoreViewSet,
                basename="threatscore")
router = DefaultRouter()
router.register(r'defendedassets',
                views_assets_tracks.DefendedAssetViewSet, basename='defendedasset')
router.register(r'tracks', views_assets_tracks.TrackViewSet, basename='track')

# --- quick ping to verify we’re editing the right file ---


def ping(_): return JsonResponse({"ok": True})


urlpatterns = [
    path("ping", ping, name="ping"),  # /api/tewa/ping -> {"ok": true}

    path("", views.root, name="root"),
    path("scenarios/", views.scenarios, name="scenarios"),
    path("score/", views.score, name="score-list-alias"),

    # compute / analytics / uploads
    # intentionally no trailing slash
    path("compute_at", views.compute_at, name="compute-at"),
    path("compute_now/", views.compute_now, name="compute_now"),
    path("ranking/", views.ranking, name="ranking"),
    path("calculate_scores/", views.calculate_scores, name="calculate_scores"),
    path("upload_tracks/", views.upload_tracks, name="upload_tracks"),

    # Task 21 — Score breakdown (both spellings)
    path("score-breakdown",  score_breakdown, name="score-breakdown"),
    path("score_breakdown/", score_breakdown, name="score_breakdown_alias"),

    # Task 22 — CSV export
    path("export/threat_board.csv", views.export_threat_board_csv,
         name="export_threat_board_csv"),

    # Task 23 — Scenario params
    path("scenarios/<int:scenario_id>/params/",
         ScenarioParamsView.as_view(), name="scenario_params"),

    # Task 24 — PNG chart
    path("charts/score_history.png",
         score_history_png_view, name="score_history_png"),
    path("score_breakdown",  score_breakdown, name="score_breakdown"),
    path("score_history.png", score_history_png_view, name="score_history_png"),
    # Add alias for v1 API
    path("v1/compute_now/", views.compute_now, name="compute_now_v1"),
    path(
        "api/threatscores/<int:scenario_id>/",
        views.api_threatscores,
        name="api_threatscores",
    ),

]

urlpatterns += router.urls
