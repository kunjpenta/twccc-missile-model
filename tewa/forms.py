# tewa/forms.py
from __future__ import annotations

from django import forms
from django.core.exceptions import ValidationError

from tewa.models import DefendedAsset, ModelParams


# ------------------------------
# Defended Asset Form
# ------------------------------
class DefendedAssetForm(forms.ModelForm):
    class Meta:
        model = DefendedAsset
        fields = ["name", "lat", "lon", "radius_km"]
        widgets = {
            "name": forms.TextInput(attrs={"class": "form-control", "placeholder": "Enter DA name"}),
            "lat": forms.NumberInput(attrs={"class": "form-control", "step": "0.000001", "placeholder": "Latitude"}),
            "lon": forms.NumberInput(attrs={"class": "form-control", "step": "0.000001", "placeholder": "Longitude"}),
            "radius_km": forms.NumberInput(attrs={"class": "form-control", "step": "0.1", "placeholder": "Radius in km"}),
        }

    def clean_radius_km(self):
        radius = self.cleaned_data.get("radius_km")
        if radius is None:
            raise ValidationError("Radius is required.")
        if radius <= 0 or radius > 1000:
            raise ValidationError("Radius must be between 0 and 1000 km.")
        return radius

    def clean_lat(self):
        lat = self.cleaned_data.get("lat")
        if lat is None:
            raise ValidationError("Latitude is required.")
        if lat < -90 or lat > 90:
            raise ValidationError(
                "Latitude must be between -90 and 90 degrees.")
        return lat

    def clean_lon(self):
        lon = self.cleaned_data.get("lon")
        if lon is None:
            raise ValidationError("Longitude is required.")
        if lon < -180 or lon > 180:
            raise ValidationError(
                "Longitude must be between -180 and 180 degrees.")
        return lon


# ------------------------------
# Track Upload Form
# ------------------------------
class UploadTrackForm(forms.Form):
    file = forms.FileField(required=True)
    scenario_id = forms.IntegerField(required=False)


# ------------------------------
# Scenario Parameters Form
# ------------------------------
class ScenarioParamsForm(forms.ModelForm):
    class Meta:
        model = ModelParams
        fields = [
            "R_W_m", "R_DA_m", "tick_s",
            "w_cpa", "w_tcpa", "w_tdb", "w_twrp",
            "sigma_cpa", "sigma_tcpa", "sigma_tdb", "sigma_twrp",
        ]
        widgets = {
            "R_W_m": forms.NumberInput(attrs={"min": 1, "step": 1}),
            "R_DA_m": forms.NumberInput(attrs={"min": 0, "step": 1}),
            "tick_s": forms.NumberInput(attrs={"min": 0.000001, "step": "any"}),
            "w_cpa": forms.NumberInput(attrs={"min": 0, "max": 1, "step": 0.01}),
            "w_tcpa": forms.NumberInput(attrs={"min": 0, "max": 1, "step": 0.01}),
            "w_tdb": forms.NumberInput(attrs={"min": 0, "max": 1, "step": 0.01}),
            "w_twrp": forms.NumberInput(attrs={"min": 0, "max": 1, "step": 0.01}),
            "sigma_cpa": forms.NumberInput(attrs={"min": 0, "step": "any"}),
            "sigma_tcpa": forms.NumberInput(attrs={"min": 0, "step": "any"}),
            "sigma_tdb": forms.NumberInput(attrs={"min": 0, "step": "any"}),
            "sigma_twrp": forms.NumberInput(attrs={"min": 0, "step": "any"}),
        }

    def clean(self):
        data = super().clean()

        # --- Weights sanity ---
        def f(x): return float(x) if x is not None else 0.0
        wsum = f(data.get("w_cpa")) + f(data.get("w_tcpa")) + \
            f(data.get("w_tdb")) + f(data.get("w_twrp"))
        if abs(wsum - 1.0) > 1e-6:
            raise forms.ValidationError("Weights must sum to 1.0")

        # --- Tick rate ---
        tick = data.get("tick_s")
        if tick is None or tick <= 0:
            self.add_error("tick_s", "Tick rate must be greater than 0")

        # --- Range consistency ---
        R_W, R_DA = data.get("R_W_m"), data.get("R_DA_m")
        if R_W is not None and R_DA is not None and R_DA >= R_W:
            self.add_error(
                "R_DA_m", "R_DA_m must be positive and less than R_W_m")

        return data
