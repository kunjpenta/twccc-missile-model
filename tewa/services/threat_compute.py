# tewa/services/threat_computes.py

from __future__ import annotations

from datetime import timezone as dt_timezone
from time import time
from typing import Any, Dict, List, Mapping, Optional, Union, cast

from django.utils import timezone

from core.utils.geodesy import LatLon, enu_from_latlon
from tewa.models import DefendedAsset, ModelParams, Scenario, ThreatScore, Track
from tewa.services import sampling
from tewa.services.kinematics import compute_cpa_tcpa_tdb_twrp
from tewa.services.normalize import clamp01, inv1
from tewa.services.scoring import _coerce_params
from tewa.services.scoring import (
    score_components_to_threat as _score_components_to_threat,
)
from tewa.types import ParamLike, ParamsLike


# ------------------------
# small helpers
# ------------------------
def _distance_km_to_da_center(da: DefendedAsset, lat: float, lon: float) -> float:
    e, n = enu_from_latlon(LatLon(lat, lon), LatLon(da.lat, da.lon))
    return (e * e + n * n) ** 0.5 / 1000.0


def _iso(dt):
    return dt.astimezone(dt_timezone.utc).isoformat().replace("+00:00", "Z")


# ------------------------
# core threat computation
# ------------------------
def compute_threat_score(
    *,
    da_lat: float,
    da_lon: float,
    da_radius_km: float,
    track_lat: float,
    track_lon: float,
    speed_mps: float,
    heading_deg: float,
    model_params: Union[Mapping[str, Any], ModelParams],
    weapon_range_km: Optional[float] = None,
) -> dict[str, float]:
    """
    Compute normalized threat score from CPA/TCPA/TDB/TWRP using parameter weights.
    Returns: {"cpa_km","tcpa_s","tdb_km","twrp_s","score"}
    """
    p = _coerce_params(cast(ParamsLike, model_params))

    bundle = compute_cpa_tcpa_tdb_twrp(
        da_lat=da_lat,
        da_lon=da_lon,
        da_radius_km=da_radius_km,
        trk_lat=track_lat,
        trk_lon=track_lon,
        speed_mps=speed_mps,
        heading_deg=heading_deg,
        weapon_range_km=(weapon_range_km or da_radius_km),
    )

    # --- normalization scales ---
    cpa_scale_km = float(p.get("cpa_scale_km", 20.0))
    tcpa_scale_s = float(p.get("tcpa_scale_s", 120.0))
    tdb_scale_km = float(p.get("tdb_scale_km", 30.0))
    twrp_scale_s = float(p.get("twrp_scale_s", 120.0))

    # Negative TCPA → past → normalize as ∞
    tcpa_norm = None if (
        bundle.tcpa_s is not None and bundle.tcpa_s < 0) else bundle.tcpa_s

    n_cpa = inv1(bundle.cpa_km, cpa_scale_km)
    n_tcpa = inv1(tcpa_norm, tcpa_scale_s)
    n_tdb = inv1(bundle.tdb_s, tdb_scale_km)
    n_twrp = inv1(bundle.twrp_s, twrp_scale_s)

    # --- weights ---
    w_cpa = float(p.get("w_cpa", 0.25))
    w_tcpa = float(p.get("w_tcpa", 0.25))
    w_tdb = float(p.get("w_tdb", 0.25))
    w_twrp = float(p.get("w_twrp", 0.25))
    clamp = bool(p.get("clamp_0_1", True))

    score = w_cpa * n_cpa + w_tcpa * n_tcpa + w_tdb * n_tdb + w_twrp * n_twrp
    if clamp:
        score = clamp01(score)

    return {
        "cpa_km": float(bundle.cpa_km),
        "tcpa_s": max(0.0, float(bundle.tcpa_s or 0.0)),
        "tdb_km": float(bundle.tdb_s),
        "twrp_s": float("inf") if bundle.twrp_s is None else float(bundle.twrp_s),
        "score": float(score),
    }


def score_components_to_threat(
    cpa_km: float,
    tcpa_s: Optional[float],
    tdb_km: float,
    twrp_s: Optional[float],
    params: ParamLike | ParamsLike | Mapping[str, Any] | ModelParams,
) -> float:
    """Wrapper to delegate to canonical scoring with type-flexible params."""
    return _score_components_to_threat(
        cpa_km=cpa_km,
        tcpa_s=tcpa_s,
        tdb_km=tdb_km,
        twrp_s=twrp_s,
        params=cast(ParamsLike, params),
    )


def compute_score_for_track(
    scenario: Scenario,
    da: DefendedAsset,
    track: Track,
    params: ParamLike | Mapping[str, Any] | ModelParams | ParamsLike,
    weapon_range_km: Optional[float] = None,
) -> ThreatScore:
    """
    Compute and persist the threat score for one track–DA pair.
    Uses normalized weights and scales, safe defaults, and full kinematic bundle.
    """
    # Coerce params (dict or ORM)
    p = _coerce_params(cast(ParamsLike, params))

    # Safe numeric access
    w_cpa = float(p.get("w_cpa", 0.0))
    w_tcpa = float(p.get("w_tcpa", 0.0))
    w_tdb = float(p.get("w_tdb", 0.0))
    w_twrp = float(p.get("w_twrp", 0.0))

    # Normalize zero weights
    if (w_cpa + w_tcpa + w_tdb + w_twrp) == 0.0:
        for k in ("w_cpa", "w_tcpa", "w_tdb", "w_twrp"):
            p[k] = 0.25

    # Compute all kinematic components
    bundle = compute_cpa_tcpa_tdb_twrp(
        da_lat=da.lat,
        da_lon=da.lon,
        da_radius_km=da.radius_km,
        trk_lat=track.lat,
        trk_lon=track.lon,
        speed_mps=track.speed_mps,
        heading_deg=track.heading_deg,
        weapon_range_km=weapon_range_km or da.radius_km,
    )

    # Compute final score
    score = score_components_to_threat(
        cpa_km=bundle.cpa_km,
        tcpa_s=bundle.tcpa_s,
        tdb_km=bundle.tdb_s,
        twrp_s=bundle.twrp_s,
        params=cast(ParamsLike, p),
    )

    # Persist ThreatScore row
    return ThreatScore.objects.create(
        scenario=scenario,
        track=track,
        da=da,
        cpa_km=bundle.cpa_km,
        tcpa_s=bundle.tcpa_s,
        tdb_km=bundle.tdb_s,
        twrp_s=bundle.twrp_s,
        score=score,
        computed_at=timezone.now(),
    )


def batch_compute_for_scenario(
    scenario_id: int,
    da_id: int,
    weapon_range_km: float | None = None,
) -> list[ThreatScore]:
    """
    Compute threat scores for all tracks in a given scenario/DA pair.
    Returns a list of ThreatScore objects.
    """
    start = time()
    scenario = Scenario.objects.get(id=scenario_id)
    da = DefendedAsset.objects.get(id=da_id)

    # Ensure parameters exist
    params, _ = ModelParams.objects.get_or_create(
        scenario=scenario,
        defaults=dict(
            w_cpa=0.25,
            w_tcpa=0.25,
            w_tdb=0.25,
            w_twrp=0.25,
            cpa_scale_km=20.0,
            tcpa_scale_s=120.0,
            tdb_scale_km=30.0,
            twrp_scale_s=120.0,
            clamp_0_1=True,
        ),
    )

    out: list[ThreatScore] = []
    for track in Track.objects.filter(scenario=scenario).iterator():
        out.append(
            compute_score_for_track(
                scenario=scenario,
                da=da,
                track=track,
                params=cast(ParamsLike, params),
                weapon_range_km=weapon_range_km,
            )
        )

    duration = round(time() - start, 3)
    print(f"[TEWA] Computed {len(out)} scores in {duration}s")
    return out


def calculate_scores_for_when(
    scenario: Scenario,
    when,
    das: list[DefendedAsset],
    method: str = "linear",
    weapon_range_km: float = 20.0,
) -> List[Dict]:
    """Pure compute (no DB writes), used by analytics or playback."""
    params_obj = (
        ModelParams.objects.filter(scenario=scenario).first()
        or ModelParams.objects.first()
    )
    if not params_obj:
        return []

    P = _coerce_params(cast(ParamsLike, params_obj))

    tracks = Track.objects.filter(scenario=scenario).only("id", "track_id")
    results: List[Dict] = []

    for tr in tracks:
        state = sampling.get_state(tr, when=when, method=method)
        if not state:
            continue

        lat, lon, spd, hdg = (
            state["lat"],
            state["lon"],
            state["speed_mps"],
            state["heading_deg"],
        )

        for da in das:
            bundle = compute_cpa_tcpa_tdb_twrp(
                da_lat=da.lat,
                da_lon=da.lon,
                da_radius_km=da.radius_km,
                trk_lat=lat,
                trk_lon=lon,
                speed_mps=spd,
                heading_deg=hdg,
                weapon_range_km=weapon_range_km or da.radius_km,
            )

            n_cpa = inv1(bundle.cpa_km, P.get("cpa_scale_km", 20.0))
            n_tcpa = inv1(
                bundle.tcpa_s if (
                    bundle.tcpa_s is None or bundle.tcpa_s >= 0) else float("inf"),
                P.get("tcpa_scale_s", 120.0),
            )
            n_tdb = inv1(bundle.tdb_s, P.get("tdb_scale_km", 30.0))
            n_twrp = inv1(
                bundle.twrp_s if (
                    bundle.twrp_s is None or bundle.twrp_s >= 0) else float("inf"),
                P.get("twrp_scale_s", 120.0),
            )

            score = (
                P.get("w_cpa", 0.25) * n_cpa
                + P.get("w_tcpa", 0.25) * n_tcpa
                + P.get("w_tdb", 0.25) * n_tdb
                + P.get("w_twrp", 0.25) * n_twrp
            )
            if P.get("clamp_0_1", True):
                score = clamp01(score)

            results.append(
                dict(
                    track_id=tr.track_id,
                    da_name=da.name,
                    score=round(float(score), 6),
                    components=dict(
                        dcpa=bundle.cpa_km,
                        tcpa=bundle.tcpa_s,
                        tdb=bundle.tdb_s,
                        twrp=bundle.twrp_s,
                        n_dcpa=n_cpa,
                        n_tcpa=n_tcpa,
                        n_tdb=n_tdb,
                        n_twrp=n_twrp,
                    ),
                    sampled_at=_iso(state["sampled_at"]),
                )
            )

    results.sort(key=lambda r: r["score"], reverse=True)
    return results
