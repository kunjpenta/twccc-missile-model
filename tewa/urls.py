# tewa/urls.py  — HTML routes only

# tewa/urls.py — HTML routes only
from django.contrib.auth import views as auth_views
from django.urls import path
from django.views.generic import TemplateView

from . import views
from .views import scenario_assumptions_view

app_name = "tewa"

urlpatterns = [
     
    path("", views.home, name="home"),
    path("about/", TemplateView.as_view(template_name="about.html"), name="about"),
    path("contact/", TemplateView.as_view(template_name="contact.html"), name="contact"),

    # Auth
    path("login/", auth_views.LoginView.as_view(template_name="registration/login.html"), name="login"),
    path("logout/", auth_views.LogoutView.as_view(), name="logout"),

    path("scenarios/<int:scenario_id>/",
         views.scenario_detail, name="scenario_detail"),
    path("scenarios/<int:scenario_id>/compute-now/",
         views.compute_now_scenario, name="compute_now_scenario"),

    path("da/", views.da_list, name="da_list"),
    path("da/create/", views.da_create, name="da_create"),
    path("da/<int:da_id>/edit/", views.da_edit, name="da_edit"),
    path("da/<int:da_id>/delete/", views.da_delete, name="da_delete"),

    path("upload-tracks/", views.upload_tracks_form, name="upload_tracks_form"),
    path("tracks/browser/", views.track_browser_page, name="track_browser_page"),

    path("score_breakdown/<int:scenario_id>/<int:da_id>/<int:track_id>/",
         views.score_breakdown_page, name="score_breakdown_page"),

    path("scenarios/<int:scenario_id>/assumptions/",
         scenario_assumptions_view, name="scenario_assumptions"),


]
