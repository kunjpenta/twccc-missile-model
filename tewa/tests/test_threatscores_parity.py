# tewa/tests/test_threatscores_parity.py
from decimal import Decimal

import pytest
from django.urls import reverse

from tewa.models import DefendedAsset, Scenario, ThreatScore, Track
from tewa.services.threat_compute import (
    batch_compute_for_scenario,  # ✅ correct import
)

pytestmark = pytest.mark.django_db


@pytest.fixture
def seeded_threatscores():
    """Seed deterministic scenario with known lat/lon coordinate schema."""
    s = Scenario.objects.create(name="Scenario-ParityTest")

    # Defended Asset (center)
    da = DefendedAsset.objects.create(
        scenario=s,
        name="DA-Test",
        lat=0.0,
        lon=0.0,
        radius_km=10.0,
    )

    # Track (slightly offset from DA)
    t = Track.objects.create(
        scenario=s,
        track_id="T-Alpha",
        lat=0.1,
        lon=0.1,
        alt_m=1000.0,           # ✅ Add altitude to satisfy NOT NULL
        speed_mps=250.0,
        heading_deg=225.0,
    )

    # Compute and persist threat scores
    from tewa.services.threat_compute import batch_compute_for_scenario
    batch_compute_for_scenario(s.id, da.id)

    return s


def _approx_equal(a, b, tol=1e-6):
    return abs(Decimal(a) - Decimal(b)) <= Decimal(str(tol))


def test_threatscores_api_matches_db(client, seeded_threatscores):
    """Verify API and DB ThreatScores parity within rounding tolerance."""
    s = seeded_threatscores
    url = reverse("tewa_api:api_threatscores", kwargs={"scenario_id": s.id})
    resp = client.get(url)
    assert resp.status_code == 200

    api_rows = resp.json()
    db_rows = list(ThreatScore.objects.filter(scenario=s).order_by("id"))

    assert len(api_rows) == len(db_rows) > 0

    for api_row, db_row in zip(api_rows, db_rows):
        for field in ["cpa_km", "tcpa_s", "tdb_km", "twrp_s", "score"]:
            a = round(float(api_row[field]), 8)
            b = round(float(getattr(db_row, field)), 8)
            assert _approx_equal(a, b), f"{field} mismatch: {a} vs {b}"

    # Confirm API sorted by score descending
    api_sorted = sorted(api_rows, key=lambda x: x["score"], reverse=True)
    assert api_rows == api_sorted


def test_snapshot_guard(client, seeded_threatscores, snapshot):
    """Guard against formula drift (snapshot of deterministic outputs)."""
    s = seeded_threatscores
    url = reverse("tewa_api:api_threatscores", kwargs={"scenario_id": s.id})
    resp = client.get(url)
    assert resp.status_code == 200
    data = resp.json()

    for row in data:
        for key in ["cpa_km", "tcpa_s", "tdb_km", "twrp_s", "score"]:
            row[key] = round(float(row[key]), 6)

    snapshot.assert_match(data)
