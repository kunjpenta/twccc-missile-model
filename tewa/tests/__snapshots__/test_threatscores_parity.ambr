# serializer version: 1
# name: test_snapshot_guard
  list([
    dict({
      'computed_at': '2025-10-16T11:41:39.384Z',
      'cpa_km': 3e-06,
      'da_id': 2,
      'id': 2,
      'scenario_id': 2,
      'score': 0.765731,
      'tcpa_s': 62.901424,
      'tdb_km': 22.901424,
      'track_id': 2,
      'twrp_s': 22.901337,
    }),
  ])
# ---
